File: frontend/src/services/billingService.ts
import type { InsuranceClaim, Invoice, Payment } from '../types/billing';
import api from './api';

export const billingService = {
  // Invoice Operations
  createInvoice: async (invoice: Partial<Invoice>): Promise<Invoice> => {
    const response = await api.post<Invoice>('/billing/invoices', invoice);
    return response.data;
  },

  getInvoice: async (id: number): Promise<Invoice> => {
    const response = await api.get<Invoice>(`/billing/invoices/${id}`);
    return response.data;
  },

  getPatientInvoices: async (patientId: number): Promise<Invoice[]> => {
    const response = await api.get<Invoice[]>(
      `/billing/invoices/patient/${patientId}`
    );
    return response.data;
  },

  getOutstandingInvoices: async (): Promise<Invoice[]> => {
    const response = await api.get<Invoice[]>('/billing/invoices/outstanding');
    return response.data;
  },

  // Payment Operations
  recordPayment: async (payment: Partial<Payment>): Promise<Payment> => {
    const response = await api.post<Payment>('/billing/payments', payment);
    return response.data;
  },

  getInvoicePayments: async (invoiceId: number): Promise<Payment[]> => {
    const response = await api.get<Payment[]>(
      `/billing/payments/invoice/${invoiceId}`
    );
    return response.data;
  },

  getPaymentReport: async (
    startDate: string,
    endDate: string
  ): Promise<Payment[]> => {
    const response = await api.get<Payment[]>('/billing/payments/report', {
      params: { start_date: startDate, end_date: endDate },
    });
    return response.data;
  },

  // Insurance Claim Operations
  submitClaim: async (
    claim: Partial<InsuranceClaim>
  ): Promise<InsuranceClaim> => {
    const response = await api.post<InsuranceClaim>('/insurance/claims', claim);
    return response.data;
  },

  getClaim: async (id: number): Promise<InsuranceClaim> => {
    const response = await api.get<InsuranceClaim>(`/insurance/claims/${id}`);
    return response.data;
  },

  getPendingClaims: async (): Promise<InsuranceClaim[]> => {
    const response = await api.get<InsuranceClaim[]>(
      '/insurance/claims/pending'
    );
    return response.data;
  },

  approveClaim: async (id: number, approvedAmount: number): Promise<void> => {
    await api.post(`/insurance/claims/${id}/approve`, {
      approved_amount: approvedAmount,
    });
  },

  rejectClaim: async (id: number, reason: string): Promise<void> => {
    await api.post(`/insurance/claims/${id}/reject`, { reason });
  },
};

File: backend/internal/service/billing_service.go
package service

import (
	"errors"
	"fmt"
	"time"

	"github.com/zarishsphere/zarish-his/internal/models"
	"github.com/zarishsphere/zarish-his/internal/repository"
)

type BillingService struct {
	repo *repository.BillingRepository
}

func NewBillingService(repo *repository.BillingRepository) *BillingService {
	return &BillingService{repo: repo}
}

func (s *BillingService) GenerateInvoice(invoice *models.Invoice) error {
	// Generate invoice number
	invoice.InvoiceNumber = s.generateInvoiceNumber()
	invoice.InvoiceDate = time.Now()

	// Set due date (30 days from invoice date)
	invoice.DueDate = invoice.InvoiceDate.AddDate(0, 0, 30)

	// Calculate item amounts
	for i := range invoice.Items {
		item := &invoice.Items[i]
		item.Amount = float64(item.Quantity) * item.UnitPrice
		item.NetAmount = item.Amount - item.Discount + item.Tax
	}

	invoice.Status = "pending"
	invoice.PaidAmount = 0

	return s.repo.CreateInvoice(invoice)
}

func (s *BillingService) GetInvoice(id uint) (*models.Invoice, error) {
	return s.repo.GetInvoice(id)
}

func (s *BillingService) GetPatientInvoices(patientID uint) ([]models.Invoice, error) {
	return s.repo.ListInvoices(patientID, "")
}

func (s *BillingService) GetOutstandingInvoices() ([]models.Invoice, error) {
	return s.repo.GetOutstandingInvoices()
}

func (s *BillingService) RecordPayment(payment *models.Payment) error {
	// Validate payment amount
	invoice, err := s.repo.GetInvoice(payment.InvoiceID)
	if err != nil {
		return err
	}

	if payment.Amount > invoice.BalanceAmount {
		return errors.New("payment amount exceeds balance")
	}

	payment.PaidAt = time.Now()
	payment.Status = "completed"

	return s.repo.CreatePayment(payment)
}

func (s *BillingService) GetInvoicePayments(invoiceID uint) ([]models.Payment, error) {
	return s.repo.GetPaymentsByInvoice(invoiceID)
}

func (s *BillingService) GetPaymentReport(startDate, endDate time.Time) ([]models.Payment, error) {
	return s.repo.GetPaymentsByDateRange(startDate, endDate)
}

func (s *BillingService) SubmitInsuranceClaim(claim *models.InsuranceClaim) error {
	// Generate claim number
	claim.ClaimNumber = s.generateClaimNumber()
	claim.SubmittedAt = time.Now()
	claim.Status = "submitted"

	return s.repo.CreateClaim(claim)
}

func (s *BillingService) GetClaim(id uint) (*models.InsuranceClaim, error) {
	return s.repo.GetClaim(id)
}

func (s *BillingService) GetPendingClaims() ([]models.InsuranceClaim, error) {
	return s.repo.ListClaims("submitted")
}

func (s *BillingService) ApproveClaim(id uint, approvedAmount float64) error {
	return s.repo.UpdateClaimStatus(id, "approved", approvedAmount, "")
}

func (s *BillingService) RejectClaim(id uint, reason string) error {
	return s.repo.UpdateClaimStatus(id, "rejected", 0, reason)
}

func (s *BillingService) MarkClaimPaid(id uint) error {
	claim, err := s.repo.GetClaim(id)
	if err != nil {
		return err
	}

	if claim.Status != "approved" {
		return errors.New("claim must be approved before marking as paid")
	}

	return s.repo.UpdateClaimStatus(id, "paid", claim.ApprovedAmount, "")
}

// Helper functions
func (s *BillingService) generateInvoiceNumber() string {
	return fmt.Sprintf("INV-%d-%d", time.Now().Unix(), time.Now().Nanosecond()%1000)
}

func (s *BillingService) generateClaimNumber() string {
	return fmt.Sprintf("CLM-%d-%d", time.Now().Unix(), time.Now().Nanosecond()%1000)
}
